### PR TITLE
Destination Redshift: handle empty s3 bucket_path

### PR DESCRIPTION
**Pull request by edgao**
discovered in https://github.com/airbytehq/oncall/issues/916

`check` will attempt to create an empty object at `s3://test_bucket/test/bucket/path/`. If you have an empty string for bucket_path, it attempts to create an object with the empty string `""` as its path, which S3 does not allow. Add code to handle this:
* In most cases, behave the same
* If s3_bucket_path is null or empty-string, then don't create a directory object
    * this also changes how we generate sub-objects, since they don't need a `/` separator between the bucket path and filename

do not publish until https://github.com/airbytehq/airbyte/pull/18031 is merged, to avoid pushing a regression to destination-redshift.

### CODE REVIEW

**Review by ryankfu — APPROVED** [review]

**Comment by ryankfu** [comment]
Changes make sense to have, what prompted this change?

**Comment by edgao** [comment]
derp, forgot to add the oc issue link to the description (https://github.com/airbytehq/oncall/issues/916)

**Comment by ryankfu** [comment]
@edgao based on [this comment](https://github.com/airbytehq/oncall/issues/916#issuecomment-1290768454) should there also be an edit to the docs saying that if `root` access is desired to merely have `/` as the `bucket_path`? 
Screen Shot 2022-10-25 at 13 11 42

**Comment by edgao** [comment]
oh hm, maybe I should figure out how to make it work then if this is supposed to be a supported thing

**Comment by edgao** [comment]
@ryankfu I rewrote the entire pr 😅  ptal again

File: airbyte-integrations/bases/base-java-s3/src/test/java/io/airbyte/integrations/destination/s3/S3BaseChecksTest.java [hunk 1]
```diff
@@ -49,4 +52,45 @@ public void attemptWriteAndDeleteS3Object_should_createSpecificFiles() {
...
+    S3BaseChecks.attemptS3WriteAndDelete(operations, config, "");
+
+    verify(s3Client, never()).putObject("test_bucket", "", "");
+    verify(s3Client).putObject(eq("test_bucket"), startsWith("_airbyte_connection_test_"), anyString());
+    verify(s3Client).listObjects(ArgumentMatchers.<ListObjectsRequest>argThat(request -> "test_bucket".equals(request.getBucketName())));
+    verify(s3Client).deleteObject(eq("test_bucket"), startsWith("_airbyte_connection_test_"));
+  }
+
+  @Test
+  public void attemptWriteAndDeleteS3Object_should_skipDirectoryCreateIfNullPath() {
```
Comment (ryankfu) [inline airbyte-integrations/bases/base-java-s3/src/test/java/io/airbyte/integrations/destination/s3/S3BaseChecksTest.java:77]: Looks good overall, just noticed that there's a seemingly random `_` in the method name and in the existing methods (so presume you're keeping with what's already existing). Minor nit but to keep test naming conventions, would it make more sense to have this consistent with camelCase?

**Review by ryankfu — APPROVED** [review]
Looks good, just a minor comment about method naming and the oddity of it in relation to all other method naming conventions in the codebase but not blocking

File: airbyte-integrations/bases/base-java-s3/src/test/java/io/airbyte/integrations/destination/s3/S3BaseChecksTest.java (hunk 1 above)
Comment (edgao) [reply airbyte-integrations/bases/base-java-s3/src/test/java/io/airbyte/integrations/destination/s3/S3BaseChecksTest.java:77]: lol this is actually my preferred testcase naming convention - imo the `testWhateverMethod` style is harder to read and doesn't provide much info about what the testcase is intended to do

iirc there's a few of these scattered through the codebase (which is where I picked it up from). Not strongly opinionated on this but mildly prefer to spread this style further.

**Pull request merged**